Improve number formats

sgiehl opened the pull request:
This PR aims to improve the number formats used by Piwik. (refs #1254)

Currently the number formats mostly depend on the server locale. So on some systems the decimal separator is a `.` where on another system it might be a `,`

With this changes number formatting data will be included in the Intl Plugin. With a new NumberFormatter in Core, as well as a plugin for jQPlot the number formatting will be done depending on the language.

The new number formatting makes a difference between numbers only and percentage values. As in some languages it is `35%` where in others it might be `% 35`. (Note: we could later use that for money formats aswell)

Places where new number formatting is currently used:
- Datatables (automatically done for all numbers)
- all jqplot graphs (axis values and tooltips)
- MetricFormatter (might affect some API results)

This branch is based on dateformats branch. So it needs a rebase after #8856 is merged

sgiehl committed:
fetch number formats and symbols from cldr

sgiehl committed:
fixed some tests after full cldr update

sgiehl committed:
added first version of numberformatter & use it for datatables

sgiehl committed:
fix php compatibility

sgiehl committed:
added number formatter for jqplot and use it for numbers in yaxis

sgiehl committed:
improve jqplot number formatter and use it for data in tooltips

sgiehl committed:
improved numberformatter, fixed tests

mattab commented on plugins/Intl/lang/id.json:
```diff
@@ -450,36 +500,72 @@
         "OneMinute": "1 menit",
         "OneMinuteShort": "1 mnt",
-        "OriginalLanguageName": "Bahasa Indonesia",
+        "OriginalLanguageName": "Indonesia",
```
this is regression in language name, see https://id.wikipedia.org/wiki/Bahasa_Indonesia

sgiehl replied:
That is what CLDR returns. They have done an update within the last month. So some strings might have changed

mattab commented:
Wow, we have number formatting, finally! Awesome added value @sgiehl 

I love it already. the UI just looks better with a simple change like this. reports become easier to read and analyse! 

Feedback

 * There are three system tests failing (added comma as thousands separators)
 * One of them introduces a No Breaking space character where maybe we should avoid it? (`U+00A0	 	\xc2\xa0	&#xA0;	  	NO-BREAK SPACE`) here:
```
-    <bounce_rate>100%</bounce_rate>
-    <exit_rate>100%</exit_rate>
+    <bounce_rate>100&#xA0;%</bounce_rate>
+    <exit_rate>100&#xA0;%</exit_rate>
```
 * Could you add a few tests for NumberFormatter?

### suggestions
 * [ ] add number formatting to scheduled reports (email at least, SMS not needed)
 * [ ] add number formatting to sparklines reports (Visitors, Referrers and Goals/Ecommerce overviews)
 * [ ] Indicate in Dev changelog any change to API output - refs #8856 `prettyDate` value has changed

sgiehl commented:
Rebased on master.

@mattab Regarding your feedback:
I've changed the MetricFormatter to use the new formatting instead of using number_format, but that changed the API responses. Imho we should have API responses in english format always (or completely unformatted). Otherwise it might get harder to work with them afterwards. You've already mentioned the example with the non breaking space in french format.

Maybe I should revert the changes to MetricFomatter and rewrite it to use no or english formats always?

I'll have a look at the other suggestions...

sgiehl committed:
added tests for NumberFormatter

sgiehl committed:
adds currency formats to intl data

mattab commented:
> Maybe I should revert the changes to MetricFomatter and rewrite it to use no or english formats always?

To make sure we don't regress API output in LTS, here is a suggestion:

 * In LTS 2.x we don't modify the API response

 * In 3.0.0 branch we can modify API response 

What do you think?

> I'll have a look at the other suggestions...

Thanks, looking forward to merge :)

mattab commented:
Hi @sgiehl it would be awesome to get number formatting before next week so we can cut RC next week :-)

sgiehl committed:
use new currency number formats

sgiehl committed:
use number formats for values displayed besides sparkline and for evolution data

sgiehl committed:
refactored js number formatter to own object to make it better usable

sgiehl committed:
use number formats in transition popover

sgiehl committed:
use number formats for row evolution popover

sgiehl committed:
use number formats in SMS/PDF reports and in static image graphs

sgiehl committed:
use number formats for SEO metrics

sgiehl committed:
use number formats in visitor map

sgiehl committed:
submodule updates

sgiehl commented:
Small update: I've added support for currency formats and added usage of number formats to scheduled reports, the numbers besides the sparklines, seo metrics, visitor map, row evolution and transition popover.
Also I've undone the changes that had an impact on the api results

mattab committed:
SMS report: percentage evolution are prefixed by '+'
https://github.com/piwik/piwik/pull/8857

mattab committed:
Use + sign from language, do not assume '+'

mattab commented:
Hi @sgiehl nice progress!

Merging the PR, will make the build green, then cut a beta and test further on demo.piwik.org
